tests/test_cqcconnection.py

```diff
@@ -0,0 +1,78 @@
+import pytest
+
+from cqc.pythonLib import CQCConnection, qubit
+from cqc.cqcHeader import CQCHeader, CQCCmdHeader, CQC_TP_COMMAND,\
+    CQC_CMD_HDR_LENGTH, CQC_CMD_H, CQC_CMD_NEW, CQC_CMD_RELEASE
+
+from utilities import get_header
+
+
+def get_expected_headers_simple_h():
+    """What headers we expect"""
+    hdr_tp_cmd = get_header(
+        CQCHeader, 
+        version=2,
+        tp=CQC_TP_COMMAND,
+        app_id=0,
+        length=CQC_CMD_HDR_LENGTH,
+    )
+    hdr_cmd_new = get_header(
+        CQCCmdHeader, 
+        qubit_id=0,
+        instr=CQC_CMD_NEW,
+        notify=True,
+        action=False,
+        block=True,
+    )
+    hdr_cmd_h = get_header(
+        CQCCmdHeader, 
+        qubit_id=0,
+        instr=CQC_CMD_H,
+        notify=True,
+        action=False,
+        block=True,
+    )
+    hdr_cmd_release = get_header(
+        CQCCmdHeader, 
+        qubit_id=0,
+        instr=CQC_CMD_RELEASE,
+        notify=True,
+        action=False,
+        block=True,
+    )
+
+    expected_headers = [
+        hdr_tp_cmd,
+        hdr_cmd_new,
+        hdr_tp_cmd,
+        hdr_cmd_h,
+        hdr_tp_cmd + hdr_cmd_release,
+    ]
+
+    return expected_headers
+
+
+def commands_to_apply_simple_h(cqc):
+    """What to do with the CQCConnection"""
+    q = qubit(cqc)
+    q.H()
+
+
+@pytest.mark.parametrize("commands_to_apply, get_expected_headers", [
+    (commands_to_apply_simple_h, get_expected_headers_simple_h),
+])
+def test_commands(commands_to_apply, get_expected_headers, monkeypatch, mock_socket, mock_read_message):
+
+    with CQCConnection("Test", socket_address=('localhost', 8000), use_classical_communication=False) as cqc:
+        commands_to_apply(cqc)
+
+    expected_headers = get_expected_headers()
+
+    commands_send = list(filter(lambda call: call.name == 'send', cqc._s.calls))
+    assert len(expected_headers) == len(commands_send)
+    for command, expected in zip(commands_send, expected_headers):
+        print(command.args[0])
+        print(expected)
+        print()
+        if expected is not None:
```

Comment:
In what situation would expected be None?

Reply:
Done